### PR TITLE
957 - etags, 404 filtering, cache control, networking updates and zoom fixes.

### PR DESCRIPTION
**Pull request by crankycoder**
This is a large patch, but I've rebased it into something that is hopefully digestable.

In order the patches are:
- d7c1924 is a straight copy of the osmdroid codebase that we've been using until now.  There are no code modifications.
- 9ae2bf0 which moves org.osmdroid into org.mozilla.osmdroid.  Code has been reformatted to force brace brackets as well as organizing imports.  No other code change is involved.
- 8721509 is where I squashed all changes to osmdroid.  

The key bits that you will care about are:
1) I've removed MapTileProvider and MapTileFilesystemProvider as we no longer use them
2) Those have been replaced with SmartFSProvider which is a facade around tile loading from disk or network.  The SmartFSProvider caches tiles on disk for 300 seconds, does a conditional GET of the tile and save the tile to disk with an etag and a cache control file (which is used to enforce the 300 second cache rule)
3) MapTileDownloader has been replaced with a TileDownloaderDelegate.  This uses HttpUrlConnection instead of the Apache DefaultHttpClient.  I've added socket timeouts and connection timeouts as well as using connection recycling if the Android API level is appropriate.
4) ZoomInLooper and ZoomOutLooper have had the handleTile method changed to a no-op.  This prevents tiles from rescaling when a zoom operation occurs.  osmdroid's zoom bug (https://github.com/osmdroid/osmdroid/issues/52) is caused (basically) because of a poor design which marks tiles as 'dirty'.  There's no easy way to fix this so I simply don't scale tiles anymore.  

There an LRUCache in play which marks tiles as 'dirty' and in need of a download, the cache somtimes evicts tiles before they can be pulled from cache and redownloaded, so I've also forced the OVERSHOOT_TILE_CACHE_SIZE to be 100 tiles (from 0) to increase the chances that the LRU doesn't evict tiles too quickly.  The LRUCache should really be removed entirely and this part of the code needs a redesign.

There are low hanging fruit CPU optimizations we can make to remove the calls to handleTile() so that we aren't looping over all tiles in a zoom just to do nothing.

5) There's also a cache that keeps HTTP404_CACHE I've embedded into the TileDownloaderDelegate which makes sure that we don't keep requesting non-existant coverage tiles from Ichnaea. 

25131a8 adds :
1) adds GET support to HttpUtil as well as providing a means to access the raw bytes of the downloaded content.
2) All Log.e() calls now require a Throwable type as the last argument.  You cannot just pass strings to a Log.e() call.  If you do, you should probably be using Log.w() for warnings.

The last two commits ccc2c65 and db98db7 are small adjustments to the build system and to reduce the logging level of some verbose and info messages down to debug as they are not important to normal runtime execution.

If you checkout this code into a separate branch, you can diff the mozstumbler changes using:

```
git diff -w dev android/src/main/java/org/mozilla/mozstumbler
```

### CODE REVIEW

**Comment by garvankeeley** [comment]
Just started looking getting this in the log:
10-22 23:08:07.147  10623-10795/org.mozilla.mozstumbler E/Stumbler:TileWriter﹕ Failed to read etag file: [/storage/emulated/0/Android/data/org.mozilla.mozstumbler/cache/osmdroid/tiles/MozStumbler Tile Store/14/4578/5974.png32.etag]
    java.io.FileNotFoundException: /storage/emulated/0/Android/data/org.mozilla.mozstumbler/cache/osmdroid/tiles/MozStumbler Tile Store/14/4578/5974.png32.etag: open failed: ENOENT (No such file or directory)
            at libcore.io.IoBridge.open(IoBridge.java:409)
            at java.io.FileInputStream.<init>(FileInputStream.java:78)
            at java.io.FileInputStream.<init>(FileInputStream.java:105)
            at org.mozilla.osmdroid.tileprovider.modules.TileWriter.readEtag(TileWriter.java:150)
            at org.mozilla.osmdroid.tileprovider.modules.TileDownloaderDelegate.isTileCurrent(TileDownloaderDelegate.java:83)
            at org.mozilla.osmdroid.tileprovider.modules.SmartFSProvider$TileLoader.loadTile(SmartFSProvider.java:183)
            at org.mozilla.osmdroid.tileprovider.modules.MapTileModuleProviderBase$TileLoader.run(MapTileModuleProviderBase.java:289)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
     Caused by: libcore.io.ErrnoException: open failed: ENOENT (No such file or directory)
            at libcore.io.Posix.open(Native Method)
            at libcore.io.BlockGuardOs.open(BlockGuardOs.java:110)
            at libcore.io.IoBridge.open(IoBridge.java:393)
            at java.io.FileInputStream.<init>(FileInputStream.java:78)
            at java.io.FileInputStream.<init>(FileInputStream.java:105)
            at org.mozilla.osmdroid.tileprovider.modules.TileWriter.readEtag(TileWriter.java:150)
            at org.mozilla.osmdroid.tileprovider.modules.TileDownloaderDelegate.isTileCurrent(TileDownloaderDelegate.java:83)
            at org.mozilla.osmdroid.tileprovider.modules.SmartFSProvider$TileLoader.loadTile(SmartFSProvider.java:183)
            at org.mozilla.osmdroid.tileprovider.modules.MapTileModuleProviderBase$TileLoader.run(MapTileModuleProviderBase.java:289)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)

**Comment by garvankeeley** [comment]
Is working well so far

**Comment by crankycoder** [comment]
Hrm.  After scrolling around a lot,  tile downloads seem to stop for me.

**Comment by Djfe** [comment]
normal OSM map tiles should have a longer cache time than 300 seconds -> they don't change too much
3 days or more should be good
for Ichnaea tiles: it could be more than just 300 seconds (I thought there was the 24h timer or is Ichnaea now creating new tiles when needed instead of doing it at midnight?)

**Comment by crankycoder** [comment]
@garvankeeley that exception is a noisey log message.  It shouldn't have been logged as an error.  I'll correct that. 

My tiles not loading problem is a bitmap related OOM problem:

```
  90 E/Stumbler:BitmapTileSourceBase( 3632): OutOfMemoryError loading bitmap: /storage/emulated/0/Android/data/org.mozilla.mozstumbler/cache/osmdroid/tiles/MozStumbler Tile Store/18/73229/95670.png32     .tile
  91 E/Stumbler:BitmapTileSourceBase( 3632): java.lang.OutOfMemoryError
  92 E/Stumbler:BitmapTileSourceBase( 3632): >---at android.graphics.BitmapFactory.nativeDecodeStream(Native Method)
  93 E/Stumbler:BitmapTileSourceBase( 3632): >---at android.graphics.BitmapFactory.decodeStreamInternal(BitmapFactory.java:620)
  94 E/Stumbler:BitmapTileSourceBase( 3632): >---at android.graphics.BitmapFactory.decodeStream(BitmapFactory.java:596)
  95 E/Stumbler:BitmapTileSourceBase( 3632): >---at android.graphics.BitmapFactory.decodeFile(BitmapFactory.java:376)
  96 E/Stumbler:BitmapTileSourceBase( 3632): >---at org.mozilla.osmdroid.tileprovider.tilesource.BitmapTileSourceBase.getDrawable(BitmapTileSourceBase.java:110)
  97 E/Stumbler:BitmapTileSourceBase( 3632): >---at org.mozilla.osmdroid.tileprovider.modules.SmartFSProvider$TileLoader.loadTile(SmartFSProvider.java:192)
  98 E/Stumbler:BitmapTileSourceBase( 3632): >---at org.mozilla.osmdroid.tileprovider.modules.MapTileModuleProviderBase$TileLoader.run(MapTileModuleProviderBase.java:289)
  99 E/Stumbler:BitmapTileSourceBase( 3632): >---at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
 100 E/Stumbler:BitmapTileSourceBase( 3632): >---at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
 101 E/Stumbler:BitmapTileSourceBase( 3632): >---at java.lang.Thread.run(Thread.java:841)

```

File: android/src/main/java/org/mozilla/mozstumbler/service/core/logging/Log.java [hunk 1]
```diff
@@ -29,19 +21,22 @@ public static void w(String logTag, String s) {
         AppGlobals.guiLogInfo(logTag + ":" + s);
     }
 
-    public static void e(String logTag, String s, Exception e) {
+    public static void e(String logTag, String s, Throwable e) {
         if (BuildConfig.BUILD_TYPE.equals("unittest")) {
             System.out.println("E: " + logTag + ", " + s);
-            e.printStackTrace();
+            if (e != null) {
```
Comment (garvankeeley) [inline android/src/main/java/org/mozilla/mozstumbler/service/core/logging/Log.java:27]: toString the exception, we only need the first line of the exception. A full stacktrace is overkill on the log.

**Comment by garvankeeley** [comment]
Works well. 
- On first zoom, the screen blanks. We will have to live with that for now though, requires too many more changes to OSMDroid. 
- Need to verify that ichnaea does not change the ETAGS on tiles with no new data.
- Test case for tile expiry would be good to have

**Pull request merged**